openedx/core/djangoapps/discussions/migrations/0001_initial.py

```diff
@@ -0,0 +1,101 @@
+# Generated by Django 2.2.16 on 2020-10-16 16:36
+
+from django.conf import settings
+from django.db import migrations, models
+import django.db.models.deletion
+import django.utils.timezone
+import jsonfield.encoder
+import jsonfield.fields
+import model_utils.fields
+import opaque_keys.edx.django.models
+import simple_history.models
+
+
+class Migration(migrations.Migration):
+
+    initial = True
+
+    dependencies = [
+        ('sites', '0002_alter_domain_unique'),
+        ('organizations', '0001_squashed_0007_historicalorganization'),
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='DiscussionProviderConfig',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('created', model_utils.fields.AutoCreatedField(default=django.utils.timezone.now, editable=False, verbose_name='created')),
+                ('modified', model_utils.fields.AutoLastModifiedField(default=django.utils.timezone.now, editable=False, verbose_name='modified')),
+                ('name', models.CharField(help_text='A user-friendly name for this configuration. e.g. SomeOrg Discourse', max_length=100)),
+                ('provider', models.CharField(db_index=True, help_text='The discussion tool/provider.', max_length=100, verbose_name='Discussion provider')),
+                ('config', jsonfield.fields.JSONField(default={}, dump_kwargs={'cls': jsonfield.encoder.JSONEncoder, 'separators': (',', ':')}, help_text='The configuration data for this provider.', load_kwargs={})),
+                ('restrict_to_org', models.ForeignKey(blank=True, help_text='Optionally restrict this config for use only on this Organization.', null=True, on_delete=django.db.models.deletion.CASCADE, to='organizations.Organization')),
```

Comment:
Just to sanity check this - we're saying that when an org or site is deleted, we should delete all discussion provider configs that referenced it.

I _think_ that makes sense... is there any situation where we would want one of these configs to survive an associated org or site?  Should the foreign key be set to null so that the config persists and goes back into the general population of configs, minus its restriction?

Reply:
I thought about this, but given the way it works, I didn't feel comfortable not cascading the deletion. If we set it to null, that will remove the restriction as you mentioned, but that may mean that a configuration (which contains secret credentials) will now become globally available. 

Even though only admins will be able to see it either way, it will become visible as an option to other course staff, which may not be desirable. Other options I considered were: add a setting to switch over to a new default org on delete, and add a signal to disable the config if the org is deleted. 

Finally though, I realised that since we're already using django simple history here, it will maintain historical records of the model even if it's deleted, so at least the data isn't lost. If a situation comes up where a delete needs to be reverted the historical record can do that.

Reply:
I think this makes sense - thanks for talking it through!